pilot/pkg/proxy/envoy/v2/ads_common.go

```diff
@@ -0,0 +1,181 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v2
+
+import (
+	"istio.io/istio/pilot/pkg/features"
+	"istio.io/istio/pilot/pkg/model"
+	"istio.io/istio/pkg/config/schemas"
+)
+
+func ProxyNeedsPush(proxy *model.Proxy, pushEv *XdsEvent) bool {
+	if !features.ScopePushes.Get() {
+		// If push scoping is not enabled, we push for all proxies
+		return true
+	}
+
+	targetNamespaces := pushEv.namespacesUpdated
+	configs := pushEv.configTypesUpdated
+
+	// appliesToProxy starts as false, we will set it to true if we encounter any configs that require a push
+	appliesToProxy := false
+	// If no config specified, this request applies to all proxies
+	if len(configs) == 0 {
+		appliesToProxy = true
+	}
+Loop:
+	for config := range configs {
+		switch config {
+		case schemas.Gateway.Type:
+			if proxy.Type == model.Router {
+				return true
+			}
+		case schemas.QuotaSpec.Type, schemas.QuotaSpecBinding.Type:
+			if proxy.Type == model.SidecarProxy {
+				return true
+			}
+		default:
+			appliesToProxy = true
+			break Loop
+		}
+	}
+
+	if !appliesToProxy {
+		return false
+	}
+
+	// If no only namespaces specified, this request applies to all proxies
+	if len(targetNamespaces) == 0 {
+		return true
+	}
+
+	// Otherwise, only apply if the egress listener will import the config present in the update
```

Comment:
Maybe a minor bug today: if you add a service that selects a pod we should send new inbound config. I think that works since they depend on themselves do maybe just comment is misleading?

Reply:
If a service selects the proxy, the egress listener should import the config, otherwise it will not get pushed.  As stated here, the logic is not right, we need to push to the proxy if its service updated.

Will fix it.